Jetpack Credentials: Refactor credentials test logic to make it reusable

Initsogar opened the pull request:
We require to prevent site restores if Jetpack credentials are not valid. For this purpose, @villanovachile made an important change in #66468, implementing a new API to test credentials on Jetpack Credentials Setting page.

The idea of this PR is to move the credentials test logic to a reusable approach across components, so it can be used in Jetpack Settings and also on Backup/Restore Flow. The implementation for Backup/Restore Flow will be handled in a separate PR.

#### Proposed Changes
* Move the API call logic to a data-layer and create the necessary actions, reducers and selectors to handle the new state.
* Refactor `AdvancedCredentials` component to use this new logic.
  * If a site has no credentials, mark them as invalid, so we don't need to call the API in those cases.
  * If we update site credentials, mark them as valid, as the update process test the credentials as part of its procedure.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
_(reusing #66468 testing instructions as this refactor impacts the same functionalities)_
* Spin up a new Jurassic Ninja site (or use an existing self hosted site), connect Jetpack, and add a Jetpack Backup plan
* Add server credentials at cloud.jetpack.com > Settings, or follow the steps here: https://jetpack.com/support/backup/backups-via-the-jetpack-plugin/adding-credentials-to-jetpack/
* Navigate to cloud.jetpack.com > Settings, or to WPCOM > Settings > Jetpack, and confirm that it shows connected (results should match test from RWDB)
* SSH into your site and change password, then navigate back to cloud.jetpack.com > Settings or to WPCOM > Settings > Jetpack, and Calypso should show Disconnected prompting user to re-enter credentials

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #66468 & #67561

Initsogar committed:
Add data-layer logic to test Jetpack credentials

Currently this is implemented on AdvancedCredentials component, the
idea is to abstract this logic so it can be reused in other components.

Initsogar committed:
Add test credentials action and reducer

Initsogar committed:
Add selector to get test credential status from state

Initsogar committed:
Refactor AdvancedCredentials to use the new credentials test action

Initsogar committed:
Trigger test success when credentials are updated

Initsogar committed:
Mark credentials as invalid when there's no credentials

This is useful as we don't need to call the API unnecessarily.

villanovachile commented:
Everything tested great on my end. I really like the fact that the test is not run if credentials are not entered as well, as this saves on unnecessary load!

Here are my testing findings:

1. Initial prompt to enter credentials as expected shows `Disconnected` since no credentials are present.

2. After correct credentials are entered, refreshing the page tests for credentials and shows `Connected`, mirroring RWDB

3. Changing password via SSH and refreshing page tests credentials and displays `Disconnected` status, mirroring RWDB and prompting to enter credentials

4. Changing password back to correct password via SSH and refreshing page test credentials and displays `Connected` status, mirroring RWDB again

5. Deleting credentials from RWDB and refreshing page immediately bypasses testing credentials and displays `Disconnected` status, mirroring RWDB, and prompting user to enter credentials 

6. Deleting credentials in Calypso, then refreshing page correctly displays `Disconnected`, without testing credentials, and mirrors RWDB, prompting user to enter credentials

tyxla commented on client/state/selectors/get-jetpack-credentials-test-status.js:
```diff
@@ -0,0 +1,7 @@
+import 'calypso/state/jetpack/init';
+
+export default function getJetpackCredentialsTestStatus( state, siteId, role ) {
+	return get( state, [ 'jetpack', 'credentials', 'testRequestStatus', siteId, role ], 'pending' );
```
I'd suggest not introducing any new Lodash usage, as we've been trying to get rid of it - see p4TIVU-9Bf-p2. I believe you should be able to get around this by using optional chaining and nullish coalescing.

Initsogar replied:
Good to know, I will work on it.

Initsogar replied:
This is now fixed.

tyxla commented on client/state/data-layer/wpcom/sites/rewind/credentials/index.js:
```diff
@@ -0,0 +1,53 @@
+			query: {
+				role: action.role,
+			},
+			body: {},
```
Is the empty `body` necessary?

Initsogar replied:
Unfortunately, it seems there's a bug on `http()` that doesn't consider `apiNamespace` for POST requests if we don't specify `body`. I wasn't able to get where's precisely the bug, but I will double-check it and let you know.

tyxla replied:
If that's the case, that bug sounds like a good opportunity to fix something on the framework side of things, which we're always happy to see.

Initsogar replied:
That's a good idea. I'll take notes about it and after this implementation we can consider taking some time and dig a bit for the root cause.

tyxla replied:
Sounds like a good idea for a follow-up 👍

tyxla commented on client/components/advanced-credentials/index.tsx:
```diff
@@ -89,26 +94,42 @@ const AdvancedCredentials: FunctionComponent< Props > = ( { action, host, role }
+				setTestCredentialsResult( false );
+			}
+		},
+		[ credentialsTestStatus ]
```
Aren't `setTestCredentialsLoading` and `setTestCredentialsResult` also dependencies here?

Initsogar replied:
We are passing the `credentialsTestStatus` to this effect, as we want to re-run it when only the `credentialsTestStatus` changes. Is there any recommendation/suggestion about why it would be likely to include those as dependency for that effect (regardless of the implementation)?

tyxla replied:
I was wrongly left with the impression that these are action creators. Since they're `useState` setters, they're fine to be omitted, as per the [React hooks docs](https://reactjs.org/docs/hooks-reference.html#usestate):

> React guarantees that `setState` function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the `useEffect` or `useCallback` dependency list.

tyxla commented on client/state/data-layer/wpcom/sites/rewind/credentials/index.js:
```diff
@@ -0,0 +1,53 @@
+	);
+
+export const testCredentialsSucceeded = ( { siteId, role }, { ok } ) => ( {
+	type: JETPACK_CREDENTIALS_TEST_SUCCESS,
```
I'd suggest making sure that any actions that are dispatched anywhere (including by the data layer) are created by action creator functions, declared in `actions.js`. This allows a single update path towards debugging and editing a particular action, and that reduces the ambiguity and the potential for errors.

Initsogar replied:
So the idea is to declare those functions on `actions.js` and call them from anywhere using something like `dispatch( testCredentials( siteId, role ) );` or `dispatch( markCredentialsAsInvalid( siteId, role ) );`, am I right?

I will be on it.

tyxla replied:
Yes 👍

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -63,6 +66,21 @@ export const updateRequestStatus = keyedReducer( 'siteId', ( state, { type } ) =
 	return state;
 } );
 
+export const testRequestStatus = keyedReducer( 'siteId', ( state, { type, role, testResult } ) => {
```
I'd suggest adding some tests to cover the reducer functionality that we're introducing here.

Initsogar replied:
I will be on it.

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -63,6 +66,21 @@ export const updateRequestStatus = keyedReducer( 'siteId', ( state, { type } ) =
+export const testRequestStatus = keyedReducer( 'siteId', ( state, { type, role, testResult } ) => {
+	switch ( type ) {
+		case JETPACK_CREDENTIALS_TEST:
+			return { [ role ]: 'pending' };
```
Won't this override any other existing keys for other roles in this reducer? Should we maintain them?

Initsogar replied:
I think this is okay. As soon as we start the credentials check process, this state will be `pending` until we get a response, then it will be `valid` or `invalid`.

tyxla replied:
In that case, since we always override the state for a site, I'd recommend a different schema, though, because the one suggested here is less predictable.

How about something like:

```
{ role: 'main', status: 'invalid' }
```

Initsogar replied:
As we discussed today, we missed to keep existing keys on that state, so we are now adding:

```
case JETPACK_CREDENTIALS_TEST_VALID:
	return {
		...state,
		[ role ]: 'valid',
	};
```

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -63,6 +66,21 @@ export const updateRequestStatus = keyedReducer( 'siteId', ( state, { type } ) =
+			return { [ role ]: 'pending' };
+
+		case JETPACK_CREDENTIALS_TEST_SUCCESS:
+			return { [ role ]: testResult ? 'valid' : 'invalid' };
```
Instead of storing a string value, should we just store a boolean `true` or `false`?

Initsogar replied:
Not possible as we need to handle 3 different states: `pending`, `valid` and `invalid`. I just omitted the `failure` state [based on this](https://github.com/Automattic/wp-calypso/pull/67589/files#r980698684).

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -63,6 +66,21 @@ export const updateRequestStatus = keyedReducer( 'siteId', ( state, { type } ) =
+			return { [ role ]: testResult ? 'valid' : 'invalid' };
+
+		case JETPACK_CREDENTIALS_TEST_FAILURE:
+			return { [ role ]: 'failed' };
```
What's the difference between failed and invalid, and what difference do they make when it comes to implementation? Can we ideally hold a single failure state?

Initsogar replied:
Well, technically the failure state is when we have an internal issue connecting to VP/Transport Server. For this implementation, I think we can consider failure as invalid because we will ask again for credentials.

Initsogar replied:
I'll make the change and let you know.

tyxla commented on client/state/selectors/get-jetpack-credentials-test-status.js:
```diff
@@ -0,0 +1,7 @@
+
+import 'calypso/state/jetpack/init';
+
+export default function getJetpackCredentialsTestStatus( state, siteId, role ) {
```
Selectors are usually also good candidates for writing some unit tests, to ensure they work well in all instances (when we have the data and when we don't).

Initsogar replied:
Tests has been added for this selector :)

Initsogar committed:
Remove Lodash usage and add optional chaining and nullish coalescing

This commit also includes testing for the getJetpackCredentialsTestStatus()
selector.

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/credentials-test

Initsogar committed:
Refactor actions to be dispatched using action creator functions

Also we are now considering only `pending`, `valid` and `invalid` credentials test state.

Initsogar committed:
Add unit test for credentials test reducer

Initsogar committed:
Remove `failure` logic for credentials test state

Initsogar commented:
Hi @tyxla:

Thanks for your help reviewing this code. Based on your comments, a made some changes on the PR and I would like you to take a look on it. :)

- Removed lodash usage and used optional chaining and nullish coalescing instead
- Refactored actions to be dispatched using action creator functions (`actions.js`)
- Added unit tests for the new selector
- Added unit tests for credentials test reducer
- Removed possible ambiguity with different test credentials status (invalid vs failure)

tyxla commented on client/state/selectors/get-jetpack-credentials-test-status.js:
```diff
@@ -0,0 +1,5 @@
+import 'calypso/state/jetpack/init';
+
+export default function getJetpackCredentialsTestStatus( state, siteId, role ) {
+	return state.jetpack?.credentials?.testRequestStatus?.[ siteId ]?.[ role ] || 'pending';
```
Given that we're loading the Jetpack reducer on the first line above, I don't think there's a scenario where `state.jetpack` will not exist, so we should be able to ditch the optional chaining there. Same for the `credentials` subreducer - if the Jetpack reducer is loaded, it always loads the `credentials` subreducer, so that should always be available. Same for the `credentials` subreducer - if the Jetpack reducer is loaded, it always loads the `testRequestStatus` subreducer, so that should always be available.

```suggestion
	return state.jetpack.credentials.testRequestStatus.[ siteId ]?.[ role ] || 'pending';
```

Note: if our tests are failing because those optional chaining instances are missing, then our tests are likely not correctly specifying the empty state scenario - again, in mind with the fact that `state.jetpack` will always be there when this selector is called.

Initsogar replied:
Thanks! It is now fixed.

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -0,0 +1,46 @@
+			expect( testRequestStatus( undefined, {} ) ).toEqual( {} );
+		} );
+
+		test( 'state should be valid if the credentials are okay', () => {
```
A little nit on the test naming here: I believe `state` here is used for `status`, and `state` in the context of the reducer means something else, thus this could create confusion for someone reading the tests.

Initsogar replied:
Fixed!

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -0,0 +1,46 @@
+		} );
+
+		test( 'state should be valid if the credentials are okay', () => {
+			const stateIn = undefined;
```
How can `stateIn` be `undefined` when the default state of the reducer is an empty object?

Initsogar replied:
You are right! Fixing this.

Initsogar replied:
Fixed!

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -0,0 +1,46 @@
+		} );
+
+		test( 'state should be invalid if the credentials are not okay', () => {
+			const stateIn = undefined;
```
I see we always test with `undefined` starting state. Should we also add some tests that cover how the reducer transitions from one non-undefined state to another?

Initsogar replied:
Good one, working on it.

Initsogar replied:
Fixed and added tests for reducer transitions!

tyxla commented on client/state/jetpack/credentials/actions.js:
```diff
@@ -48,3 +52,29 @@ export const deleteCredentials = ( siteId, role ) => ( dispatch ) => {
+	role,
+} );
+
+export const updateCredentialsTestResult = ( siteId, role, testResult ) => ( dispatch ) => {
```
If we use this thunk just once, I think it's an unnecessary indirection and can be inlined directly.

Initsogar replied:
Thanks for the suggestion. I just removed this function. In case we will need it in the future for multiple scenarios, then we will add it back.

Initsogar committed:
Fix reducer to keep existing data on testRequestStatus state

If we try to save status for multiple roles, it will now keep the previous roles statuses
on the state. Also we are adding a small fix removing unnecesary optional chaining and
adding more use cases for reducer unit tests.

Initsogar committed:
Remove unnecessary function updateCredentialsTestResult

tyxla commented on client/state/jetpack/credentials/reducer.js:
```diff
@@ -63,6 +66,30 @@ export const updateRequestStatus = keyedReducer( 'siteId', ( state, { type } ) =
+		case JETPACK_CREDENTIALS_TEST:
+			return {
+				...state,
+				[ role ]: 'pending',
```
Another idea for improvement here is to define constants with possible states like `const JETPACK_CREDENTIALS_PENDING = 'pending'` and reusing that everywhere. While that seems like a very slight improvement, it can help with reducing the potential for errors and could make the code more intuitive. In cases where those are used a lot, it can also help with code organization.

tyxla commented on client/state/jetpack/credentials/test/reducer.js:
```diff
@@ -0,0 +1,99 @@
+import { testRequestStatus } from '../reducer';
+
+describe( 'reducer', () => {
+	describe( 'testRequestStatus', () => {
```
Definitely not a blocker for this PR, but ideally, this reducer should have tests for the rest of its subreducers.

tyxla reviewed (approved):
Code is looking great to me 👍  Nice work out there!

I've left a couple of optional suggestions for follow-up. 

I haven't re-tested, so, I'd love it if we can get another round of testing to ensure this works as expected before we ship it - cc @villanovachile.

villanovachile reviewed (commented):
Thanks for the tag! All my previous testing results are replicated here as well:

>Everything tested great on my end. I really like the fact that the test is not run if credentials are not entered as well, as this saves on unnecessary load!
>
>Here are my testing findings:
>
>Initial prompt to enter credentials as expected shows Disconnected since no credentials are present.
>
>After correct credentials are entered, refreshing the page tests for credentials and shows Connected, mirroring RWDB
>
>Changing password via SSH and refreshing page tests credentials and displays Disconnected status, mirroring RWDB and prompting to enter credentials
>
>Changing password back to correct password via SSH and refreshing page test credentials and displays Connected status, mirroring RWDB again
>
>Deleting credentials from RWDB and refreshing page immediately bypasses testing credentials and displays Disconnected status, mirroring RWDB, and prompting user to enter credentials
>
>Deleting credentials in Calypso, then refreshing page correctly displays Disconnected, without testing credentials, and mirrors RWDB, prompting user to enter credentials

One thing I noted, however, is that sometimes when valid credentials are entered (maybe 1 out of 4 or 5 tries), the "Disconnect" screen will display prematurely before eventually resulting to "Connected".  I am not 100% certain what might be causing this, but it's definitely not something that I experienced in the first review. 

Here is a video demonstrating this:


https://user-images.githubusercontent.com/65082164/193139266-5ef1e996-20a5-4e2c-9d79-d8153d7e9f5a.mp4

Initsogar commented:
> #66468

Thanks for your tests and details provided @villanovachile! After reviewing the mentioned bug, it seems that is only happening on Calypso Blue occasionally. I will proceed with this PR as it is a dependency for one specific project that need this functionality and I will track this bug to fix it there or in another PR on a very short-term.
